src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java

```diff
@@ -155,7 +155,7 @@ final SQLException checkException(SQLException sqle)
       for (int depth = 0; delegate != ClosedConnection.CLOSED_CONNECTION && nse != null && depth < 10; depth++) {
          final var sqlState = nse.getSQLState();
          if (sqlState != null && sqlState.startsWith("08")
-             || nse instanceof SQLTimeoutException
+             || (nse instanceof SQLTimeoutException && nse.getSQLState().equals("HY008") && nse.getErrorCode() == 0)
```

Comment:
`nse.getSQLState()` could be replaced with `sqlState`

Also since `&&` has higher precedence than `||` I think `sqlState` could possibly be null, so the following should be safer:

`(nse instanceof SQLTimeoutException && "HY008".equals(sqlState) && nse.getErrorCode() == 0)`

Reply:
Theoretically possible, and trivial to guard against.